ash/src/entry.rs

```diff
@@ -194,21 +237,45 @@ impl<L> EntryCustom<L> {
             lib,
         })
     }
-}
 
-impl Entry {
-    pub fn new() -> Result<Self, LoadingError> {
-        Self::new_custom(
-            || {
-                DynamicLibrary::open(Some(&Path::new(LIB_PATH)))
-                    .map_err(|err| LoadingError::LibraryLoadError(err.clone()))
-                    .map(|dl| Arc::new(dl))
-            },
-            |vk_lib, name| unsafe {
-                vk_lib
-                    .symbol(&*name.to_string_lossy())
-                    .unwrap_or(ptr::null_mut())
-            },
-        )
+    #[doc = "<https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/vkEnumerateInstanceVersion.html>"]
+    /// ```rust,no_run
+    /// # #[macro_use]
+    /// # extern crate ash;
+    /// # use ash::Entry;
+    /// # fn main() -> Result<(), Box<std::error::Error>> {
+    /// let entry = Entry::new()?;
+    /// match entry.try_enumerate_instance_version()? {
+    ///     // Vulkan 1.1+
+    ///     Some(version) => {
+    ///         let major = vk_version_major!(version);
+    ///         let minor = vk_version_minor!(version);
+    ///         let patch = vk_version_patch!(version);
+    ///     },
+    ///     // Vulkan 1.0
+    ///     None => {},
+    /// }
+    /// # Ok(()) }
+    /// ```
+    pub fn try_enumerate_instance_version(&self) -> VkResult<Option<u32>> {
```

Comment:
I think we should just add this to the `Entry1_1` trait, as a convenience function. The function pointer is already in there.

The type signature would also be swapped I think `Option<VkResult<u32>`, makes more sense to me.

Reply:
I thought about removing the `Entry1_1` trait but thought it might be too dramatic :) It's not actually implemented *anywhere*. Even if it were implemented, it's not safe to call because it will panic for Vulkan 1.0 drivers.

I specifically used  `VkResult<Option<u32>>` so that `?` syntax would work. This is also aligned with other `try_` style functions in the standard library (they always return `Result`).

Reply:
`?` also works with option. We should still implement `Entry1_1`, I probably just forgot to add an `impl`. It doesn't really matter if it safe to call. No >1.0 functions are. Although this is a bit of a special case and I think a convenience function that checks if it is available seems reasonable.

I think `try_xxx` is still okay for options, `Try` is implemented for both Option and Result, unless the style guide says something more specific about this.

The question is what is more convenient for the user?

Reply:
`?` works with `Option`in functions returning `Option`. Functions returning an error type constructible from `vk::Result` will not be able to make much use out of `?` with a function returning `Option`, so I think the current strategy makes sense.

This is a bit academic, because the function is documented as having no error cases.

Reply:
Yeah I guess it doesn't really make much of a difference. 

```
ctx.try_enumerate_instance_version().map_err(...)?.unwrap_or(1.0);
```

Reply:
Actually, why not make a `EntryV1_0` function that just does that internally? If the function's missing it's by definition 1.0, so this simplifies the interface without losing any expressiveness.

Reply:
>I think try_xxx is still okay for options, Try is implemented for both Option and Result, unless the style guide says something more specific about this.

[NoneError](https://doc.rust-lang.org/std/option/struct.NoneError.html) isn't stabilized. As @Ralith stated, you can't user `?` on `Option` if the function returns a `Result`.

Reply:
@Ralith it probably also includes the patch version. Not sure if that is ever useful but its trivial to implement for the user anyway. I just used `1.0` above but it is actually an `u32`.

As I said either approach is fine by me, but 1.1 functions should stay in `1_1`, this is most likely where the user expects them to be. You also can only default implement them in `1_1`.

I would also be okay for `try_enumerate_instance_version` to just be a method as it is now, just reuse the function pointer that was already loaded and implement the `Entry1_1` trait.

Reply:
>just reuse the function pointer that was already loaded and implement the Entry1_1 trait.

It needs to check with `get_instance_proc_addr` because the `EntryFnV1_0` implementation stores a dummy function that panics (i.e. it will never be null).

Reply:
Sorry I forgot about the panic functions. I guess this is fine then?